src/query/src/datafusion.rs

```diff
@@ -568,6 +569,16 @@ impl QueryEngine for DatafusionQueryEngine {
                     });
             }
         }
+
+        state
+            .config_mut()
+            .options_mut()
+            .extensions
+            .insert(FunctionContext {
+                query_ctx: query_ctx.clone(),
+                state: self.engine_state().function_state(),
+            });
```

Comment:
we have similar code at line 529, confirm it's not duplicated

Reply:
The two are not quite the same, though they looked similar. Here is the place where `FunctionContext` can be passed down to UDF args.